devtools/datasette/publish.py

```diff
@@ -82,9 +83,75 @@ def inspect_data(datasets: list[str], pudl_out: Path) -> str:
     return inspect_output
 
 
-def metadata(pudl_out: Path) -> str:
+def check_tables_have_metadata(
+    metadata: DatasetteMetadata,
+    pudl_out: Path,
+    databases: list[str],
+) -> None:
+    """Check to make sure all tables in the databases have metadata.
+
+    This function fails if there are tables present in a database
+    that do not have structure metadata.
+
+    Args:
+        metadata: The structure metadata for the datasette deployment
+        pudl_out: The directory that contains the pudl outputs
+        databases: The list of databases to test.
+    """
+    resources = {}
+    resources |= metadata.xbrl_resources
+    resources["pudl"] = metadata.resources
+
+    database_table_exceptions = {"pudl": {"alembic_version"}}
+
+    tables_missing_metadata_results = {}
+
+    for database in databases:
+        database_path = pudl_out / database
+        database_name = database_path.stem
+
+        # Grab all tables in the database
+        engine = sa.create_engine(f"sqlite:///{str(database_path)}")
+        inspector = sa.inspect(engine)
+        tables_in_database = set(inspector.get_table_names())
```

Comment:
This is a fairly large new entanglement -- previously we did not need to have the actual databases to generate the `metadata.yml`. Now it can **only** be produced if you've already run the ETL and have the DBs laying around. Of course that's fine for the actual deployment, but is it really necessary? It will make testing or generating the metadata in any other non-deployment environment annoying or impossible.

Reply:
That makes sense. I can move check_tables_have_metadata()` so that `publish.py` can create metadata without actually needing complete databases.